File: implementations/typescript/src/eval/mobility.ts
import { Board } from "../board";
import { Color, PieceType, Square } from "../types";

const KNIGHT_MOBILITY: number[] = [-15, -5, 0, 5, 10, 15, 20, 22, 24];
const BISHOP_MOBILITY: number[] = [-20, -10, 0, 5, 10, 15, 18, 21, 24, 26, 28, 30, 32, 34];
const ROOK_MOBILITY: number[] = [-15, -8, 0, 3, 6, 9, 12, 14, 16, 18, 20, 22, 24, 26, 28];
const QUEEN_MOBILITY: number[] = [
  -10, -5, 0, 2, 4, 6, 8, 10, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 22, 23, 23, 24, 24, 25, 25, 26, 26
];

export function evaluate(board: Board): number {
  let score = 0;
  
  for (let square = 0; square < 64; square++) {
    const piece = board.getPiece(square);
    if (piece) {
      let mobility = 0;
      switch (piece.type) {
        case "N":
          mobility = countKnightMobility(board, square, piece.color);
          break;
        case "B":
          mobility = countBishopMobility(board, square, piece.color);
          break;
        case "R":
          mobility = countRookMobility(board, square, piece.color);
          break;
        case "Q":
          mobility = countQueenMobility(board, square, piece.color);
          break;
        default:
          continue;
      }
      
      const bonus = getMobilityBonus(piece.type, mobility);
      score += piece.color === "white" ? bonus : -bonus;
    }
  }
  
  return score;
}

function countKnightMobility(board: Board, square: Square, _color: Color): number {
  const offsets: [number, number][] = [
    [-2, -1], [-2, 1], [-1, -2], [-1, 2],
    [1, -2], [1, 2], [2, -1], [2, 1],
  ];
  
  const rank = Math.floor(square / 8);
  const file = square % 8;
  let count = 0;
  
  for (const [dr, df] of offsets) {
    const newRank = rank + dr;
    const newFile = file + df;
    
    if (newRank >= 0 && newRank < 8 && newFile >= 0 && newFile < 8) {
      const target = newRank * 8 + newFile;
      const targetPiece = board.getPiece(target);
      const sourcePiece = board.getPiece(square);
      
      if (targetPiece) {
        if (sourcePiece && targetPiece.color !== sourcePiece.color) {
          count++;
        }
      } else {
        count++;
      }
    }
  }
  
  return count;
}

function countBishopMobility(board: Board, square: Square, color: Color): number {
  return countSlidingMobility(board, square, color, [[1, 1], [1, -1], [-1, 1], [-1, -1]]);
}

function countRookMobility(board: Board, square: Square, color: Color): number {
  return countSlidingMobility(board, square, color, [[0, 1], [0, -1], [1, 0], [-1, 0]]);
}

function countQueenMobility(board: Board, square: Square, color: Color): number {
  return countSlidingMobility(board, square, color, [
    [0, 1], [0, -1], [1, 0], [-1, 0],
    [1, 1], [1, -1], [-1, 1], [-1, -1],
  ]);
}

function countSlidingMobility(board: Board, square: Square, color: Color, directions: number[][]): number {
  const rank = Math.floor(square / 8);
  const file = square % 8;
  let count = 0;
  
  for (const [dr, df] of directions) {
    let currentRank = rank + dr;
    let currentFile = file + df;
    
    while (currentRank >= 0 && currentRank < 8 && currentFile >= 0 && currentFile < 8) {
      const target = currentRank * 8 + currentFile;
      const targetPiece = board.getPiece(target);
      
      if (targetPiece) {
        if (targetPiece.color !== color) {
          count++;
        }
        break;
      } else {
        count++;
      }
      
      currentRank += dr;
      currentFile += df;
    }
  }
  
  return count;
}

function getMobilityBonus(pieceType: PieceType, mobility: number): number {
  switch (pieceType) {
    case "N": return KNIGHT_MOBILITY[Math.min(mobility, 8)];
    case "B": return BISHOP_MOBILITY[Math.min(mobility, 13)];
    case "R": return ROOK_MOBILITY[Math.min(mobility, 14)];
    case "Q": return QUEEN_MOBILITY[Math.min(mobility, 27)];
    default: return 0;
  }
}

File: implementations/rust/src/eval/mobility.rs
use crate::board::Board;
use crate::types::{Color, PieceType, Square};
use crate::move_generator::MoveGenerator;

const KNIGHT_MOBILITY: [i32; 9] = [-15, -5, 0, 5, 10, 15, 20, 22, 24];
const BISHOP_MOBILITY: [i32; 14] = [-20, -10, 0, 5, 10, 15, 18, 21, 24, 26, 28, 30, 32, 34];
const ROOK_MOBILITY: [i32; 15] = [-15, -8, 0, 3, 6, 9, 12, 14, 16, 18, 20, 22, 24, 26, 28];
const QUEEN_MOBILITY: [i32; 28] = [
    -10, -5, 0, 2, 4, 6, 8, 10, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 22, 23, 23, 24, 24, 25, 25, 26, 26
];

pub fn evaluate(board: &Board) -> i32 {
    let mut score = 0;
    let mg = MoveGenerator::new();
    
    for square in 0..64 {
        if let Some(piece) = board.get_piece(square) {
            let mobility = match piece.piece_type {
                PieceType::Knight => count_knight_mobility(board, square, piece.color),
                PieceType::Bishop => count_bishop_mobility(board, square, piece.color),
                PieceType::Rook => count_rook_mobility(board, square, piece.color),
                PieceType::Queen => count_queen_mobility(board, square, piece.color),
                _ => continue,
            };
            
            let bonus = get_mobility_bonus(piece.piece_type, mobility);
            score += if piece.color == Color::White { bonus } else { -bonus };
        }
    }
    
    score
}

fn count_knight_mobility(board: &Board, square: Square, _color: Color) -> usize {
    let offsets: [(i32, i32); 8] = [
        (-2, -1), (-2, 1), (-1, -2), (-1, 2),
        (1, -2), (1, 2), (2, -1), (2, 1),
    ];
    
    let rank = (square / 8) as i32;
    let file = (square % 8) as i32;
    let mut count = 0;
    
    for (dr, df) in offsets.iter() {
        let new_rank = rank + dr;
        let new_file = file + df;
        
        if new_rank >= 0 && new_rank < 8 && new_file >= 0 && new_file < 8 {
            let target = ((new_rank * 8 + new_file) as usize);
            if let Some(target_piece) = board.get_piece(target) {
                if target_piece.color != board.get_piece(square).unwrap().color {
                    count += 1;
                }
            } else {
                count += 1;
            }
        }
    }
    
    count
}

fn count_bishop_mobility(board: &Board, square: Square, color: Color) -> usize {
    count_sliding_mobility(board, square, color, &[(1, 1), (1, -1), (-1, 1), (-1, -1)])
}

fn count_rook_mobility(board: &Board, square: Square, color: Color) -> usize {
    count_sliding_mobility(board, square, color, &[(0, 1), (0, -1), (1, 0), (-1, 0)])
}

fn count_queen_mobility(board: &Board, square: Square, color: Color) -> usize {
    count_sliding_mobility(board, square, color, &[
        (0, 1), (0, -1), (1, 0), (-1, 0),
        (1, 1), (1, -1), (-1, 1), (-1, -1),
    ])
}

fn count_sliding_mobility(board: &Board, square: Square, color: Color, directions: &[(i32, i32)]) -> usize {
    let rank = (square / 8) as i32;
    let file = (square % 8) as i32;
    let mut count = 0;
    
    for (dr, df) in directions.iter() {
        let mut current_rank = rank + dr;
        let mut current_file = file + df;
        
        while current_rank >= 0 && current_rank < 8 && current_file >= 0 && current_file < 8 {
            let target = ((current_rank * 8 + current_file) as usize);
            
            if let Some(target_piece) = board.get_piece(target) {
                if target_piece.color != color {
                    count += 1;
                }
                break;
            } else {
                count += 1;
            }
            
            current_rank += dr;
            current_file += df;
        }
    }
    
    count
}

fn get_mobility_bonus(piece_type: PieceType, mobility: usize) -> i32 {
    match piece_type {
        PieceType::Knight => KNIGHT_MOBILITY[mobility.min(8)],
        PieceType::Bishop => BISHOP_MOBILITY[mobility.min(13)],
        PieceType::Rook => ROOK_MOBILITY[mobility.min(14)],
        PieceType::Queen => QUEEN_MOBILITY[mobility.min(27)],
        _ => 0,
    }
}
